runner/command.go

```diff
@@ -51,10 +57,46 @@ func (c *Command) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	return marshal.UnmarshalOneOf(doCandidate, commandCandidate)
 }
 
+// newCmd creates an exec.Cmd that uses the interpreter and the script passed
+// in a temp file to run that script with that interpreter.
+//
+// A cleanup function is returned to allow removal of the temporary file.
+func newCmd(ctx Context, script string) (*exec.Cmd, func(), error) {
+	file, err := ioutil.TempFile("", "tusk_exec")
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanup := func() {
+		os.Remove(file.Name()) // nolint: errcheck
+	}
```

Comment:
While creating a temp file is a nice _fallback_, I am not sure it's the right default. E.g. there could be security reasons for not wanting to create a temp file, such as when having a plain-text password (extracted from some vault) present on the task command-line from some variable.

It might be good (in my opinion), to:
1. If the interpreter argument for running without a temp file is _known_, use that.
2. If not, fallback to creating a temporary file.
3. State it in the documentation which interpreters are that don't generate temporary files.

This way anyone that has a security concern in relation to creating temporary files can take the right precaution.

I am far from an security expert, so I do not know if this makes any sense... Just a thought.

Reply:
It's a good thought, I'm going to have to think on this for a bit.

I lean toward being consistent over being comprehensive, so if temporary files are an issue, it probably means either supporting a whitelist of interpreters or only supporting the `${SHELL} -c "..."` pattern.

Reply:
PS! I don't _know_ if temporary files is an issue, or _when_ they are an issue.

In the example, one could imagine shells explicitly extracting "secrets" from the environment to prevent them being included in temp files. But that requires the _knowledge_ that tusk creates temporary files, which means the risk of mistakes may be higher.

From a better safe then sorry approach, then a white-list sounds like a nice place to start. If wrting to temporary files proves not to be a real problem, going from an explicit whitelist like a write-to-file general solution would be a backwards compatible change. Going the other way would not be.

Reply:
However, keep in mind there is a _a lot of shells_ out there... `bash`, `fish`, `dash`++

Not saying it's important to support them all.

One idea would be to allow two configuration formats...

A "quick" format (either allowing interpreters white-listed by tusk (map lookup), or simply assuming `args: ["-c"]` and `use_temp_file: false`):

```yaml
interpreter: python
```
A detailed format where you can set any interpreter you like:

```yaml
interpreter:
   command: "ruby"
   args: ["-e"]
```
```yaml
interpreter:
   command: "gosh"
   use_temp_file: true
```

Then it becomes very obvious when you are doing what.

Reply:
Again, this is something that can be added _later_ of starting out with a white-list.

Reply:
After thinking and talking it over with a couple folks, I don't think using a temporary file is worth the trade-offs here. The difference is only that we can't look like shebangs, which is superficial, and we can't support interpreters that can't take a literal string as a flag but can take a file, which doesn't describe any scripting language I can think of.

What's making sense to me is a configurable option that defaults to `sh -c`, but makes no assumptions about `-c` or other flags. Specified with one of these syntaxes, most likely:

1. ```yaml
    interpreter: python -c
    ```

1. ```yaml
    interpreter: [python, -c]
    ```
1. ```yaml
    interpreter:
      executable: python # maybe `path:`, maybe `command:`
      args: [-c]
    ```

We can always do something like start with option 1 or 2, then make it the short-hand version of a longer configuration like 3 down the line. It's certainly not without precedent for tusk.

A white-list that makes any assumptions (e.g., `ruby` means `ruby -e`) is something I want to stay away from. Once tusk supports it, we're stuck with a list of special interpreters until we make a breaking change.

Reply:
Or maybe the string can be passed to STDIN? Seam to work for a number of shells without having to specify any `-c`, `-e` etc.

```sh
$ echo 'print("Hello world")' | python3                                              :(
Hello world
```
```sh
$ echo 'echo "Hello world"' | sh
Hello world
```
```sh
echo 'print("Hello world\n")' | ruby
Hello world
```

Reply:
I don't have a set-up for it, but would be nice to find out if this can also work for `cmd.exe` (Batch) in Windows?

Reply:
Using stdin is something I played around with briefly, but unfortunately it means you can't read the end-user's stdin. So you wouldn't be able to do things like prompt for input or open a REPL.

Reply:
I do zero windows development, so windows support has mostly been guess work and making sure Appveyor continues to pass. I can't speak to `cmd.exe` in particular.

Reply:
Good point about STDIN. From some research, `cmd.exe` is the terminal application, so never mind that! The executable for PowerShell is `PowerShell` or `PowerShell.exe`, and it seams to allow a `-Command` option.

This means your suggested solution should work, also for that shell 👍

Docs:
- https://docs.microsoft.com/en-us/powershell/scripting/components/console/powershell.exe-command-line-help?view=powershell-7
- https://docs.microsoft.com/en-us/powershell/module/Microsoft.PowerShell.Core/About/about_PowerShell_exe?view=powershell-5.1

Besides `bash` is available on new Windows environments these days.